src/Common/HTTPConnectionPool.cpp

```diff
@@ -337,6 +381,19 @@ class EndpointConnectionPool : public std::enable_shared_from_this<EndpointConne
         std::ostream & sendRequest(Poco::Net::HTTPRequest & request) override
         {
             auto idle = idleTime();
+
+            // Reset data hooks for IO scheduling
+            if (ResourceLink link = CurrentThread::getReadResourceLink()) {
+                Session::setReceiveDataHooks(std::make_shared<ResourceGuardSessionDataHooks>(ResourceGuard::Metrics::getIORead(), link));
+            } else {
+                Session::setReceiveDataHooks();
+            }
+            if (ResourceLink link = CurrentThread::getWriteResourceLink()) {
+                Session::setSendDataHooks(std::make_shared<ResourceGuardSessionDataHooks>(ResourceGuard::Metrics::getIOWrite(), link));
+            } else {
+                Session::setSendDataHooks();
+            }
```

Comment:
I reset callbacks here because if we do that when we release the session, there might be a reader still using the stream. Is this correct or could we find a better place? Hooks are not holding any smart pointers to important objects, only a raw pointer, but when the query is destroyed it could lead to a dangling pointer. Though we should never use it, it feels not 100% safe.